ecs/pod_creator.go

```
@@ -27,11 +31,227 @@ func NewBasicECSPodCreator(c cocoa.ECSClient, v cocoa.Vault) (*BasicECSPodCreato
+	return []*ecs.PlacementStrategy{&placementStrat}
+}
+
+// exportEnvVars translates the environment variables into ECS environment variables and secrets.
+func (m *BasicECSPodCreator) exportEnvVars(ctx context.Context, variables []cocoa.EnvironmentVariable) ([]*ecs.KeyValuePair, []cocoa.PodSecret, []cocoa.NamedSecret, error) {
+	keyValuePairs := []*ecs.KeyValuePair{}
+	secrets := []cocoa.PodSecret{}
+	createSecrets := []cocoa.NamedSecret{}
+
+	for _, variable := range variables {
+		if variable.SecretOpts == nil {
+			keyValue := ecs.KeyValuePair{}
+			keyValue.SetName(*variable.Name).SetValue(*variable.Value)
+			keyValuePairs = append(keyValuePairs, &keyValue)
+		} else {
+			if !*variable.SecretOpts.Exists {
+				createSecrets = append(createSecrets, variable.SecretOpts.PodSecret.NamedSecret)
+			}
+			secrets = append(secrets, variable.SecretOpts.PodSecret)
+		}
+	}
+
+	return keyValuePairs, secrets, createSecrets, nil
+}
+
+// exportSecrets extracts all secrets and environment variables from ECS creation options
+func (m *BasicECSPodCreator) exportSecrets(ctx context.Context, merged cocoa.ECSPodCreationOptions) ([]*ecs.KeyValuePair, []cocoa.PodSecret, []cocoa.NamedSecret, error) {
+	allEnvVars := []*ecs.KeyValuePair{}
+	allSecrets := []cocoa.PodSecret{}
+	allSecretsToCreate := []cocoa.NamedSecret{}
+
+	for _, def := range merged.ContainerDefinitions {
+		envVars, secrets, secretsToCreate, err := m.exportEnvVars(ctx, def.EnvVars)
+		if err != nil {
+			return nil, nil, nil, err
+		}
+		allEnvVars = append(allEnvVars, envVars...)
+		allSecrets = append(allSecrets, secrets...)
+		allSecretsToCreate = append(allSecretsToCreate, secretsToCreate...)
+	}
+
+	return allEnvVars, allSecrets, allSecretsToCreate, nil
+}
+
+// createSecrets creates secrets that do not already exist
+func (m *BasicECSPodCreator) createSecrets(ctx context.Context, secrets []cocoa.NamedSecret) error {
+	for _, secret := range secrets {
+		_, err := m.vault.CreateSecret(ctx, secret)
```

You have to check if `m.vault` is non-nil because it's optional. If it's nil, you should probably return an error.